### PR TITLE
distribution: fix matching amd64 variants

### PR DESCRIPTION
**Pull request by tonistiigi**
These commits correct the support for pulling images with custom amd64 variants https://en.wikipedia.org/wiki/X86-64#Microarchitecture_levels that are used by the latest Go, GCC, LLVM etc.

The first commit matches variant comparison to containerd.

Second commit ensures that when no specific platform is set then the variant with maximum compatibility with the current system is used. Warning is also fixed so that it doesn't show up when variants are compatible.

Note that builder part default platform is intentionally not updated as `linux/amd64` (implicit v1) remains default there. When building new images default is to build v1 - maximum compatibility with the rest of the systems. When pulling default is to pull image with maximum compatibility for the current system.

You can use these images to test the behavior https://hub.docker.com/r/tonistiigi/test/tags?page=1&name=amd64

### CODE REVIEW

**Comment by thaJeztah** [comment]
Linter is failing;

```
distribution/pull_v2.go:1032:1: exported function `MaximumSpec` should have comment or be unexported (golint)
 func MaximumSpec() specs.Platform {
 ^

File: distribution/pull_v2.go [hunk 1]
```diff
@@ -1027,3 +1028,11 @@ func toOCIPlatform(p manifestlist.PlatformSpec) specs.Platform {
 		OSVersion:    p.OSVersion,
 	}
 }
+
```
Comment (thaJeztah) [inline distribution/pull_v2.go:1031]: should this be included in upstream's default spec? if so, we should consider un-exporting this function, or adding a note in the comment that it's temporary / deprecated

Comment (tonistiigi) [reply distribution/pull_v2.go:1031]: It should not replace `DefaultSpec` as they have different use cases. I can try to see what they think about an additional method. In https://github.com/containerd/containerd/pull/6455#issuecomment-1016696041 seemed like they wanted it, although containerd itself does not really have a use-case for this function atm.

Comment (thaJeztah) [reply distribution/pull_v2.go:1031]: I guess alternatively, github.com/tonistiigi/go-archvariant could have a helper for this, that way we could still use the same function in both places, without introducing something new in `distribution`, especially if it's (possibly) temporary (I'm trying to trim down the exported "api" of the distribution package)

Would that work for you?

Comment (tonistiigi) [reply distribution/pull_v2.go:1031]: Ideally, I'd like to not add dependencies to `go-archvariant` or expand its scope.

Comment (thaJeztah) [reply distribution/pull_v2.go:1031]: as it's (hopefully) temporary; perhaps we should just copy it to both places and make it private

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by thaJeztah** [comment]
@AkihiroSuda @cpuguy83 ptal

**Review by tianon — APPROVED** [review]

**Comment by thaJeztah** [comment]
CI is happy, so I'm happy; let's get this in

**Pull request merged**